### ISSUE TITLE
Copying data from results grid to clipboard takes minutes

### ISSUE THREAD

**Issue by OzBob** [microsoft/azuredatastudio#24901]
Type: <b>Bug</b>

simple select on azure SQL DB.
click a 'cell'
press ctrl-C
popup 'coying data...'
wait 10 seconds
paste  -  get old clipboard
wait 60 seconds
popup changes to 'copied'
past - get new text from clipboard

Azure Data Studio version: azuredatastudio 1.46.1 (ba29842b81dec01177415e53948ca2168e69c3f8, 2023-10-02T18:14:22.887Z)
OS version: Windows_NT x64 10.0.22621
Restricted Mode: No
Preview Features: Enabled
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|13th Gen Intel(R) Core(TM) i9-13900KF (32 x 2995)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|31.82GB (9.92GB free)|
|Process Argv||
|Screen Reader|no|
|VM|33%|
</details><details><summary>Extensions (8)</summary>

Extension|Author (truncated)|Version
---|---|---
sqlops-combine-scripts|Bat|2.0.1
copilot|Git|1.95.239
dacpac|Mic|1.14.0
profiler|Mic|0.14.0
schema-compare|Mic|1.21.0
sql-database-projects|Mic|1.3.1
extra-sql-script-as|pac|0.5.0
vscodeintellicode|Vis|1.2.1

</details>

**Comment by cheenamalhotra** [comment]
It shouldn't take that long. 

- What kind of data are you trying to copy? 
- Can you reproduce it consistently?
- Please gather and upload logs we can investigate.

**Comment by cheenamalhotra** [comment]
Can you provide additional info:
- How long did the query take to execute that you're copying results from?
- Did you copy results instantly after execution, or editor/IDE was left open for more than 5-10 mins before copying?

Logs would be helpful too, if you can provide.

**Comment by JohnHallstrom** [comment]
Execution time for query is 2 seconds, with 20.120 rows as a result. Copying all rows takes 1m 23s and was done as soon as the query was done executing. 

I can only provide logs that don't hold any kind of personal data or information. If you list the log files that don't contain that I can send them to you!

**Comment by serhat359** [comment]
I have the same problem. I made some local refactoring to make it much faster. I wanted to create a pull request for it but I was unable to push my branch. If someone can guide on what to do I can create a PR.

**Comment by serhat359** [comment]
In case someone wants to create a PR, the change is a one file change and it's `src/sql/workbench/services/query/common/gridDataProvider.ts`. The problem is caused by using `reduce` function on a string array (`rowValues.reduce`) and I just changed that part into creating an array, pushing the elements into the array and doing `.join("")` at the end.

**Comment by cheenamalhotra** [comment]
@serhat359 You can fork this repository and create a new branch in the fork and open PR with your change.
We are happy to welcome contributions, if you face a challenge - let us know!

**Comment by serhat359** [comment]
I created the PR https://github.com/microsoft/azuredatastudio/pull/25152. It's currently awaiting approval for running the workflows.

**Comment by kisantia** [comment]
closing issue since the improvement was merged. Thanks for the contribution!

**Comment by OzBob** [comment]
@serhat359 could you do some benchmark testing and share a rough estimate on the % improvement in transferring data into the Windows clipboard?

**Comment by serhat359** [comment]
@OzBob In the PR https://github.com/microsoft/azuredatastudio/pull/25152 I have some benchmarks, you can take a look at those. The performance improvement depends on the number of rows being copied into the clipboard. At low number of rows the gain is not really noticeable but at 10,000 rows and 50,000 rows we see 450x and 3,000x improvement respectively.

**Comment by serhat359** [comment]
I just realized the performance gain seems to happen only when the setting "Prefer Providers Copy Handler" is off. When this setting is on, ADS uses the copy handler from the provider which performs bad when the number of rows is high. Internally it might have the same problem. I did some debugging to figure out the issue and it might be caused by the package `azdata` but I'm not entirely sure. But I guess I'll be turning the setting off for the time being.
